services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/utils.py

```diff
@@ -107,29 +115,88 @@ async def write_to_tmp_file(file_contents: str) -> AsyncGenerator[Path, None]:
         await aiofiles_os.remove(file_path)
 
 
-async def async_command(
-    command: str, command_timeout: Optional[float] = None
-) -> CommandResult:
-    try:
-        proc = await asyncio.create_subprocess_shell(
-            command,
-            stdin=asyncio.subprocess.PIPE,
-            stdout=asyncio.subprocess.PIPE,
-            stderr=asyncio.subprocess.STDOUT,
-        )
+def _close_transport(proc: Process):
+
+    # Closes transport (initialized during 'await proc.communicate(...)' ) and avoids error:
+    #
+    # Exception ignored in: <function BaseSubprocessTransport.__del__ at 0x7f871d0c7e50>
+    # Traceback (most recent call last):
+    #   File " ... .pyenv/versions/3.9.12/lib/python3.9/asyncio/base_subprocess.py", line 126, in __del__
+    #     self.close()
+    #
 
-        stdout, _ = await asyncio.wait_for(proc.communicate(), timeout=command_timeout)
+    # SEE implementation of asyncio.subprocess.Process._read_stream(...)
+    for fd in (1, 2):
+        # pylint: disable=protected-access
+        if transport := getattr(proc, "_transport", None):
+            if t := transport.get_pipe_transport(fd):
+                t.close()
+
+
+async def async_command(command: str, timeout: Optional[float] = None) -> CommandResult:
+    """
+    Does not raise Exception
+    """
+    proc = await asyncio.create_subprocess_shell(
+        command,
+        stdout=asyncio.subprocess.PIPE,
+        stderr=asyncio.subprocess.STDOUT,  # TODO: might want to keep separate?
+    )
+    assert psutil.pid_exists(proc.pid)  # nosec
+
+    start = time.time()
+
+    try:
+        stdout, _ = await asyncio.wait_for(proc.communicate(), timeout=timeout)
 
     except asyncio.TimeoutError:
+        # NOTE: Sending SIGTERM to the process and awaits for termination
+        # i.e. giving the opportunity to the underying process to
+        # perform graceful shutdown (i.e. run shutdown events and cleanup tasks)
+        proc.terminate()
+        _close_transport(proc)
+
+        # There is a small risk that the process refused to shutdown
+        # when received a SIGTERM. For the moment, we will accept that risk
+        # and only check during development ...
+        #
+        assert await proc.wait() == -signal.SIGTERM  # nosec
+        assert not psutil.pid_exists(proc.pid)  # nosec
+
         logger.warning(
-            "%s timed out after %ss",
+            "Process %s timed out after %ss",
+            f"{command=!r}",
+            f"{timeout=}",
+        )
+        return CommandResult(
+            success=False,
+            message=f"Execution timed out after {timeout} secs",
+            command=f"{command}",
+            elapsed=time.time() - start,
+        )
```

Comment:
I'd guess there is no way to get the stdout at this point. Even if the command timed out?

Reply:
It is indeed possible but will complicate much more the code. I propose to leave it as is and come back if it is really needed